### PR TITLE
Antonpis/tests

### PR DESCRIPTION
**Pull request by abpiskunov**
This time i am using this PR for review latest changes to be submited to master branch. Changes that i made 2 weeks back are 
- perf optimizations to avoid extra DTE access and cache project info which used to determine if missing package suggestion should be displayed or not
- add assembly version attributes (at this moment i build this project in our CI machine and increment assembly version - let me know when/if i should use nuget's CI machine to do that)
- Bug fix: avoid adding duplicated namespaces if there is some error in package installation.

Note: there no unit tests at this moment and i am adding them. However there are E2E integration tests running in WTE builds and covering most scenarios. In WTE we use mstest for unit tests (which is convenient since it gives us codecoverage), What test driver is preferred for nuget team? xunit or mstest? If xunit how do you collect test coverage?

### CODE REVIEW

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ internal class ProjectTargetFrameworkProviderExporter : IProjectTargetFrameworkP
         /// <summary>
         /// Stores file and it's corresponding unique project name
         /// </summary>
-        private Dictionary<string, string> FilesCache = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+        private Dictionary<string, string> FilesCache = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: Why are you using InvariantCulture vs. Ordinal?

Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: This should be ordinal , I probably switched form other repo where we used Invariant all the time. I will update to Ordinal in all places in this repo.

**Comment by yishaigalatzer** [comment]
A few basic comments
1. Do not checkin binaries into a git repository.
2. Make a single PR per single change, don't mix in multiple bugs, even better make first a refactoring/cleanup change and then follow up with a code change. It makes the PR process super easy, and make the history make sense.

**Comment by abpiskunov** [comment]
I had to checkin binaries because of breaking changes in Nuget and Roslyn. If I used nuget packages - they would have changes mismatching binaries in our WPT branch and nothing would work. What solution you suggest for that?

**Comment by abpiskunov** [comment]
I already had those changes before that's why I added them all as single PR - did not have another option.

**Comment by yishaigalatzer** [comment]
For binaries you can use nuget packages, just specify an exact version.

For changes, you can always break them out to individual commits.

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs [hunk 2]
```diff
@@ -175,6 +176,16 @@ private void OnAfterSolutionClosing()
             ClearCache();
         }
 
+        private void OnProjectRemoved(Project project)
+        {
+            if (project == null)
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: when is project == null? Seems odd. Can you add a comment?

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs [hunk 3]
```diff
@@ -175,6 +176,16 @@ private void OnAfterSolutionClosing()
...
+            {
+                return;
+            }
+
+            OnProjectTargetFrameworkChanged(this, project.FullName);
+        }
+
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: I haven't see a destructor in years (it took me a few minutes to remember the spelling :) )

File: src/Nuget.PackageIndex.VisualStudio/CodeFixes/CSharp/CSharpAddPackageCodeFixProvider.cs [hunk 4]
```diff
@@ -10,6 +10,9 @@
 using Nuget.PackageIndex.Client.CodeFixes;
 using Nuget.PackageIndex.VisualStudio.Analyzers;
 using Nuget.PackageIndex.VisualStudio.CodeFixes.CSharp.Utilities;
+using System.Linq;
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/CodeFixes/CSharp/CSharpAddPackageCodeFixProvider.cs]: you want system. showing up first (it's a vs setting)

File: src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs [hunk 5]
```diff
@@ -11,24 +12,43 @@ namespace Nuget.PackageIndex.VisualStudio
...
+                {
+                    // if there is index build task already running, remember only last 
+                    // synchronize request and continue after current rebuild is complete.
+                    // This is to avoid resync 10 times when there multiple synchronize 
+                    // requests come and we don't need actually to rebuild index for them 
+                    // all, but just for the first and maybe the last one
+                    _indexBuildTask.ContinueWith(t => RebuildIndex());
+                }
             }
             catch (Exception e)
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs]: You want to log exceptions rather than debug.writeline.

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs [hunk 6]
```diff
@@ -1,20 +1,63 @@
 ﻿using System;
 using System.Diagnostics;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: sort

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs [hunk 7]
```diff
@@ -1,20 +1,63 @@
...
 using Nuget.PackageIndex.Client.Analyzers;
 
 namespace Nuget.PackageIndex.VisualStudio
 {
     /// <summary>
-    /// TODO remove this hack after RC
+    /// This filter is needed only until we enable Missing Package suggestions for all C# projects.
+    /// For now we limit this feature only to ProjectK scenarios.
     /// </summary>
     internal class ProjectKFilter : IProjectFilter
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Don't use projectK as terminology (DNX Project maybe?)

File: src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs (hunk 5 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs]: Where nuget assemblies log exceptions? I can do it, I just left it for the future.

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 7 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Good point, will change.

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs [hunk 8]
```diff
@@ -43,8 +86,114 @@ public bool IsProjectSupported(string filePath)
                     // TODO Package manager console?
                     Debug.Write(string.Format("{0} \r\n {1}", e.Message, e.StackTrace));
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: See the other comment around `Debug.Write`, in NuGet we write such errors to the output. @deepakaravindr might be able to point you to the code.

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs [hunk 9]
```diff
@@ -1,20 +1,63 @@
...
         public bool IsProjectSupported(string filePath)
         {
-            return ThreadHelper.JoinableTaskFactory.Run(async delegate
+            if (string.IsNullOrEmpty(filePath) || _solutionIsClosing)
+            {
+                return false;
+            }
+
+            string projectFullName;
+            if (FilesCache.TryGetValue(filePath, out projectFullName))
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: What if a file appears in more than one project (through links)?
Are you handling project renaming?

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs (hunk 2 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: well we receive it from the event - where it said that it must be non null? :) Just to be safe ...

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs [hunk 10]
```diff
@@ -43,8 +86,114 @@ public bool IsProjectSupported(string filePath)
                     // TODO Package manager console?
                     Debug.Write(string.Format("{0} \r\n {1}", e.Message, e.StackTrace));
                 }
-                return false;
-            });            
+            });
+
+            if (!string.IsNullOrEmpty(projectFullName))
+            {
+                lock (_cacheLock)
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: use concurrentdictionary instead of using locks

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 8 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Just to confirm - we are ok for user to see actual exceptions and stack trace in the output? We just tried to avoid that in WTE since it would look weird...

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs [hunk 11]
```diff
@@ -43,8 +86,114 @@ public bool IsProjectSupported(string filePath)
...
+
+            ClearCache(projectFullPath);
+        }
+
+        private void OnProjectRenamed(Project project, string oldName)
+        {
+            ClearCache(oldName);
+        }
+
+        ~ProjectKFilter()
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: no destructors

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 9 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Project renaming - yes line 164

File: src/Nuget.PackageIndex/ILocalPackageIndexBuilder.cs [hunk 12]
```diff
@@ -9,8 +10,8 @@ namespace Nuget.PackageIndex
     public interface ILocalPackageIndexBuilder
     {
         ILocalPackageIndex Index { get; }
-        IEnumerable<string> GetPackages(bool newOnly);
-        Task<LocalPackageIndexBuilderResult> BuildAsync(bool newOnly = false);
+        IEnumerable<string> GetPackages(bool newOnly, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex/ILocalPackageIndexBuilder.cs]: `CancellationToken.None` and avoid optional parameters on interfaces

Comment (yishaigalatzer) [inline src/Nuget.PackageIndex/ILocalPackageIndexBuilder.cs]: use extension methods to add overloads

File: src/Nuget.PackageIndex/LocalPackageIndexBuilder.cs [hunk 13]
```diff
@@ -7,6 +7,7 @@
 using System.Threading.Tasks;
 using ILog = Nuget.PackageIndex.Logging.ILog;
 using Nuget.PackageIndex.Engine;
+using System.Threading;
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex/LocalPackageIndexBuilder.cs]: sort

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 9 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: About links .. not sure yet what to do with them. Need to open an issue to track that

**Comment by yishaigalatzer** [comment]
:wait:

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs (hunk 2 above)
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: I doubt this event ever passes null, and if it does its probably a bug we should handle because something weird is going one.

File: src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs (hunk 5 above)
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs]: @deepakaravindr can help with a pointer to that. The point is that if something goes wrong here, you will never know and will never get a report either. it's just not going to work correctly.=, which is why you always log such problems.

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 9 above)
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: cool on renaming. Links is a tricky thing, many projects don't deal with it properly. Is the problem here is that for different projects you might get different recommendation?

Please indicate the issue # on the PR

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 8 above)
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Well if you have an issue, you want to share it so it can get reported.

Also instead of  `e.Message` and `e.StackTrace` and formatting you can just do `e.ToString()` same effect less electrons on the screen :)

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs (hunk 1 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: Done, using Ordinal everywhere

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs (hunk 2 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: I added Debug.Assert for us to see it internally in debug builds, but still prefer to keep null check since if it happens on users machine we would never know, but users would see null exception dialog which is not good. (I checked in index , other subscribers also check for null in such event handlers)

File: src/Nuget.PackageIndex.VisualStudio/CodeFixes/CSharp/CSharpAddPackageCodeFixProvider.cs (hunk 4 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/CodeFixes/CSharp/CSharpAddPackageCodeFixProvider.cs]: Done

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs (hunk 3 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: Fixed

File: src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs (hunk 5 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/LocalNugetPackageIndex.cs]: Opened issue https://github.com/NuGet/NuGet.PackageIndex/issues/9

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 6 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: done

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 7 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Fixed

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 9 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: The problem here is that we want to show code fix suggestions only for Dnx projects, but checking project for a file using DTE objects is time consuming since need switch to UI thread. To make it faster I added caching logic , but it might nit work correctly with linked files if one file belong to Dnx project and other one to normal Windows Console app for example. In this case depending which project we cached we would show or not show suggestions for some projects. But this should not be a big deal since, we would enable suggestions for all types of projects soon (maybe post RTM though) and this whole filter class would go away. I opened issue: https://github.com/NuGet/NuGet.PackageIndex/issues/7 to track it.

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 8 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Opened issue https://github.com/NuGet/NuGet.PackageIndex/issues/9  to track this

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 11 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Fixed

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 10 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: done

File: src/Nuget.PackageIndex/LocalPackageIndexBuilder.cs (hunk 13 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex/LocalPackageIndexBuilder.cs]: Done

File: src/Nuget.PackageIndex/ILocalPackageIndexBuilder.cs (hunk 12 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex/ILocalPackageIndexBuilder.cs]: Fixed

File: src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs (hunk 9 above)
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/ProjectKFilter.cs]: Switching to UI thread is not necessarily expensive. Did you measure to see that caching is needed?

File: src/Nuget.PackageIndex.Manager/PackageIndexManager.cs [hunk 14]
```diff
@@ -72,7 +71,7 @@ private void ProcessAction(ILocalPackageIndexBuilder builder, Arguments argument
             switch (arguments.Action)
             {
                 case PackageIndexManagerAction.Build:
-                    result = builder.BuildAsync().Result;
+                    result = builder.BuildAsync(newOnly:false, cancellationToken: CancellationToken.None).Result;
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.Manager/PackageIndexManager.cs]: you always want to auto format before saving. This will prevent unnecessary changes due to formatting on the next commit

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs [hunk 15]
```diff
@@ -104,56 +105,56 @@ public IEnumerable<TargetFrameworkMetadata> GetTargetFrameworks(string filePath)
...
                 // here we need to remove from cache files that belong to a given project
-                if (ProjectFrameworksCache.Keys.Contains(project.UniqueName))
+                if (ProjectFrameworksCache.Keys.Contains(projectFullPath))
                 {
-                    ProjectFrameworksCache.Remove(project.UniqueName);
+                    ProjectFrameworksCache.Remove(projectFullPath);
                 }
 
-                var filesToRemove = FilesCache.Where(x => x.Value.Equals(project.UniqueName, StringComparison.InvariantCultureIgnoreCase)).Select(x => x.Key).ToList();
+                var filesToRemove = FilesCache.Where(x => x.Value.Equals(projectFullPath, StringComparison.OrdinalIgnoreCase)).Select(x => x.Key).ToList();
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: Another small note - Because this code goes on github, its always a good idea to break long line. A good rule of thumb is that if you can't see the end of a line in the PR you want to break it.

File: src/Nuget.PackageIndex.VisualStudio/CodeFixes/PackageInstaller.cs [hunk 16]
```diff
@@ -137,17 +137,20 @@ private void OnPackageInstalled(IVsPackageMetadata metadata)
...
+        #region IDisposable
+
+        public void Dispose()
         {
-            Dispose();
+            Dispose(true);
+            GC.SuppressFinalize(this);
         }
 
-        public void Dispose()
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/CodeFixes/PackageInstaller.cs]: more of a question, make sure you actually want to handle this through dispose. And test your code in shutdown scenarios.

Typically you want to use dispose only when you have to get rid of native resources. It might be the right thing in your case though I'm not sure. Just a good thing to look at before picking the pattern

File: src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        private ConcurrentDictionary<string, string> FilesCache = new ConcurrentDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+        private bool _disposed;
+        private SolutionEvents _solutionEvents;
+        private object _cacheLock = new object();
+        private bool _solutionIsClosing;
+
+        public DnxProjectFilter()
+        {
+            var container = ServiceProvider.GlobalProvider.GetService<IComponentModel, SComponentModel>();
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs]: Is this guaranteed to run on the UI thread?

File: src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+            {
+                FilesCache.AddOrUpdate(filePath, (k) => projectFullName, (k, v) => projectFullName);
+            }
+
+            return IsProjectFileSupported(projectFullName);
+        }
+
+        private bool IsProjectFileSupported(string projectFile)
+        {
+            return !string.IsNullOrEmpty(projectFile) && projectFile.EndsWith(".xproj");
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs]: you probably want to rely on something less prone to change than the project extension

File: src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+
+        #region IDisposable
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Dispose(bool disposing)
```
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs]: the private method is unnecessary. The dispose(true) pattern is used for derived classes.

Also avoid `#region` in the code

**Comment by yishaigalatzer** [comment]
I added some comments, please work with billhie or someone from your team on them and signoff from your team as well.

:shipit: from me

File: src/Nuget.PackageIndex.Manager/PackageIndexManager.cs (hunk 14 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.Manager/PackageIndexManager.cs]: Sure will do

File: src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs (hunk 15 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/Analyzers/ProjectTargetFrameworkProviderExporter.cs]: Sure, will do

File: src/Nuget.PackageIndex.VisualStudio/CodeFixes/PackageInstaller.cs (hunk 16 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/CodeFixes/PackageInstaller.cs]: It is tested already for all scenarios, I just wanted to unsubscribe for events explicitly, this is how it is done In many places , but I agree it looks a bit strange. I just did not see a special place where it is safe to unsubscribe from those events. I will check with Bill if he can suggest something

File: src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs (hunk 17 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs]: It must be on UI thread (by Roslyn), but you are right, we have a helper to throw and I will add it there

File: src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs (hunk 18 above)
Comment (abpiskunov) [inline src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs]: well ..., it is temporary and we use same pattern in other places too, so it should be good enough (if extension changes it will be changed here too)

File: src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs (hunk 17 above)
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs]: An alternative is to move initialization off the constructor. In some cases its the right move, not convinced if it is necessary here. Just a thought.

-----Original Message-----

File: src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs (hunk 18 above)
Comment (yishaigalatzer) [inline src/Nuget.PackageIndex.VisualStudio/DnxProjectFilter.cs]: Work with bill if you can find a more Stable pattern that doesn't require changing things I many places

-----Original Message-----

**Comment by abpiskunov** [comment]
Merged to master. About unsubscribing in Dispose, I am checking with Roslyn when analyzers are disposed an how they clean up resources. Once I got the best place to unsubscribe will switch to it.
